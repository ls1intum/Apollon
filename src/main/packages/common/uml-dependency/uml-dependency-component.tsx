import React, { FunctionComponent } from 'react';
import { UMLDependency } from './uml-component-dependency';
import { ThemedPath, ThemedPolyline } from '../../../components/theme/themedComponents';

export const UMLDependencyComponent: FunctionComponent<Props> = ({ element }) => (
  <g>
    <marker
      id={`marker-${element.id}`}
      viewBox="0 0 30 30"
      markerWidth="22"
      markerHeight="30"
      refX="30"
      refY="15"
      orient="auto"
      markerUnits="strokeWidth"
    >
      <ThemedPath d="M0,29 L30,15 L0,1" fillColor="none" strokeColor={element.strokeColor} />
    </marker>
    <ThemedPolyline
      points={element.path.map((point) => `${point.x} ${point.y}`).join(',')}
      strokeColor={element.strokeColor}
      fillColor="none"
      strokeWidth={1}
      strokeDasharray={7}
      markerEnd={`url(#marker-${element.id})`}
    />
  </g>
);

interface Props {
  element: UMLDependency;
}
