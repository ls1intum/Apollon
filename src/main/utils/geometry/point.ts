export interface IPoint {
  x: number;
  y: number;
}

export class Point implements IPoint {
  constructor(
    public x: number = 0,
    public y: number = 0,
  ) {}

  get length(): number {
    return Math.sqrt(this.x * this.x + this.y * this.y);
  }

  normalize(): Point {
    return new Point(this.x / this.length, this.y / this.length);
  }

  add(x: number, y: number): Point;
  add(x: Point): Point;
  add(x: number | Point, y?: number): Point {
    if (x instanceof Point) {
      return new Point(this.x + x.x, this.y + x.y);
    }
    if (typeof y === 'number') {
      return new Point(this.x + x, this.y + y);
    }
    return this.clone();
  }

  subtract(x: number, y: number): Point;
  subtract(x: Point): Point;
  subtract(x: number | Point, y?: number): Point {
    if (x instanceof Point) {
      return new Point(this.x - x.x, this.y - x.y);
    }
    if (typeof y === 'number') {
      return new Point(this.x - x, this.y - y);
    }
    return this.clone();
  }

  round(radix: number = 10): Point {
    return new Point(Math.round(this.x / radix) * radix, Math.round(this.y / radix) * radix);
  }

  scale(factor: number) {
    return new Point(this.x * factor, this.y * factor);
  }

  clone() {
    return new Point(this.x, this.y);
  }
}
