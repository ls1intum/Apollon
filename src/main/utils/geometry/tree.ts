import { UMLContainer } from '../../services/uml-container/uml-container';
import { IUMLElement, UMLElement } from '../../services/uml-element/uml-element';
import { UMLElementState } from '../../services/uml-element/uml-element-types';

export function filterRoots(ids: string[], elements: UMLElementState): string[] {
  const getSelection = (root: IUMLElement): string[] => {
    if (ids.includes(root.id)) return [root.id];

    if (UMLContainer.isUMLContainer(root)) {
      return root.ownedElements.reduce<string[]>((selection, id) => [...selection, ...getSelection(elements[id])], []);
    }
    return [];
  };

  return Object.values(elements)
    .filter((element) => !element.owner)
    .reduce<string[]>((selection, element) => [...selection, ...getSelection(element)], []);
}

/**
 * returns the ids of all elements in the hierarchy above the element with the specified id
 * @param id element id which parents should be found
 * @param elements elements in state
 */
export function getAllParents(id: string, elements: UMLElementState): string[] {
  const getParents = (element: IUMLElement): string[] => {
    // reached top
    if (element.owner === null) return [];
    return [element.owner, ...getParents(elements[element.owner])];
  };
  return getParents(elements[id]);
}

export function getChildren(ids: string[], elements: UMLElementState): string[] {
  const result: string[] = [];

  for (const id of ids) {
    const owner = elements[id];
    if (!owner) continue;

    if (UMLContainer.isUMLContainer(owner)) {
      result.push(...getChildren(owner.ownedElements, elements));
    }
    result.push(owner.id);
  }

  return result;
}

export function clone(element: UMLElement, elements: UMLElement[]): UMLElement[] {
  if (!UMLContainer.isUMLContainer(element)) {
    return [element.clone()];
  }

  const result: UMLElement[] = [];
  const cloned = element.clone<UMLContainer>();
  const { ownedElements } = element;
  for (const id of ownedElements) {
    const child = elements.find((prev) => prev.id === id);
    if (!child) {
      continue;
    }

    const [clonedChild, ...clonedChildren] = clone(child, elements);
    clonedChild.owner = cloned.id;

    const index = cloned.ownedElements.findIndex((x) => x === id);
    cloned.ownedElements[index] = clonedChild.id;
    result.push(clonedChild, ...clonedChildren);
  }

  return [cloned, ...result];
}
