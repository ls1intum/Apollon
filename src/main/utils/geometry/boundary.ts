import { Point } from './point';

export interface IBoundary {
  x: number;
  y: number;
  width: number;
  height: number;
}

export function computeBoundingBox(points: Point[]): IBoundary {
  if (points.length === 0) {
    return { x: 0, y: 0, width: 0, height: 0 };
  }

  const firstPoint = points[0];

  let minX = firstPoint.x;
  let minY = firstPoint.y;

  let maxX = firstPoint.x;
  let maxY = firstPoint.y;

  for (const p of points) {
    if (p.x < minX) minX = p.x;
    if (p.y < minY) minY = p.y;

    if (p.x > maxX) maxX = p.x;
    if (p.y > maxY) maxY = p.y;
  }

  return {
    x: minX,
    y: minY,
    width: maxX - minX,
    height: maxY - minY,
  };
}

export function computeBoundingBoxForElements(elements: { bounds: IBoundary }[]): IBoundary {
  if (!elements.length) {
    return { x: 0, y: 0, width: 0, height: 0 };
  }
  const boundaries: IBoundary[] = elements.map<IBoundary>((element) => ({ ...element.bounds }));
  const x = Math.min(...boundaries.map((bounds) => bounds.x));
  const y = Math.min(...boundaries.map((bounds) => bounds.y));
  const width = Math.max(...boundaries.map((bounds) => bounds.x + bounds.width)) - x;
  const height = Math.max(...boundaries.map((bounds) => bounds.y + bounds.height)) - y;
  return { x, y, width, height };
}

export function computeDimension(scale: number, value: number, isCircle?: boolean): number {
  if (isCircle && scale === 1) {
    return value * scale;
  } else {
    return Math.round((value * scale) / 10) * 10;
  }
}
