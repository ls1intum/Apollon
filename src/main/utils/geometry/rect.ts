import { Direction } from '../../services/uml-element/uml-element-port';
import { IBoundary } from './boundary';
import { Point } from './point';

export function position(bounds: IBoundary, direction: Direction): { point: Point; offset: Point } {
  const { x, y, width, height } = bounds;
  const offset = 40;
  switch (direction) {
    case Direction.Up:
      return {
        point: new Point(x + width, y),
        offset: new Point(x + width, y - offset),
      };
    case Direction.Right:
      return {
        point: new Point(x + width, y + height),
        offset: new Point(x + width + offset, y + height),
      };
    case Direction.Down:
      return {
        point: new Point(x + width, y + height),
        offset: new Point(x + width, y + height + offset),
      };
    case Direction.Left:
      return {
        point: new Point(x, y + height),
        offset: new Point(x - offset, y + height),
      };
  }
}
