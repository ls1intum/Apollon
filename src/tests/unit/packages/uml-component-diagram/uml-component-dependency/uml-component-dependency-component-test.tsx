import * as React from 'react';
import { wrappedRender } from '../../../test-utils/render';
import { UMLDependencyComponent } from '../../../../../main/packages/common/uml-dependency/uml-dependency-component';
import { Point } from '../../../../../main/utils/geometry/point';
import { UMLComponentDependency } from '../../../../../main/packages/uml-component-diagram/uml-component-dependency/uml-component-dependency';

it('render the uml-component-dependency-component', () => {
  const dependency: UMLComponentDependency = new UMLComponentDependency({
    id: '77f00db7-ab65-41fb-9e87-aa123c7705c0',
    path: [new Point(0, 0), new Point(100, 100)],
  });
  const { getByText, baseElement } = wrappedRender(
    <svg>
      <UMLDependencyComponent element={dependency} />
    </svg>,
  );
  // TODO: expect
  // expect(getByText(dependency.name)).toBeInTheDocument();
  expect(baseElement).toMatchSnapshot();
});
