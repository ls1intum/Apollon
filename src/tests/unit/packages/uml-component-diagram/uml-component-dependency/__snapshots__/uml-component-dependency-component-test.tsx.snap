// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render the uml-component-dependency-component 1`] = `
<body>
  <div>
    <svg>
      <g>
        <marker
          id="marker-4dcd06c9-10b8-4069-9ade-7fcf6d77d6ea"
          markerHeight="30"
          markerUnits="strokeWidth"
          markerWidth="22"
          orient="auto"
          refX="30"
          refY="15"
          viewBox="0 0 30 30"
        >
          <path
            class="sc-gEvEer hATzAk"
            d="M0,29 L30,15 L0,1"
            fill="none"
            stroke="black"
          />
        </marker>
        <polyline
          class="sc-aXZVg OFiXI"
          fill="none"
          marker-end="url(#marker-4dcd06c9-10b8-4069-9ade-7fcf6d77d6ea)"
          points="0 0,100 100"
          stroke="black"
          stroke-dasharray="7"
          stroke-width="1"
        />
      </g>
    </svg>
  </div>
</body>
`;
